nitro-protocol/contracts/CurrencySwapApp.sol

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+pragma experimental ABIEncoderV2;
+
+import './interfaces/IForceMoveApp.sol';
+import './libraries/NitroUtils.sol';
+import './interfaces/INitroTypes.sol';
+import {ExitFormat as Outcome} from '@statechannels/exit-format/contracts/ExitFormat.sol';
+
+/**
+ * @dev The CurrencySwapApp contract complies with the ForceMoveApp interface and allows payments to be made virtually from Alice to Bob (participants[0] to participants[n+1], where n is the number of intermediaries).
+ */
+contract CurrencySwapApp is IForceMoveApp {
+
+    struct VoucherSwap {
```

Comment:
> * b. Maintain the invariant (off chain) that only one voucher is ever doubly signed

The problem with this is that we can only do one swap per channel, which undermines the performance benefits of virtual channels with respect to an HTLC approach (we would need to involve the intermediaries in every swap).

Reply:
1. Noted
2. Aren't your payment vouchers single use ?
In our case amount can be partial, like matching half of the resting order, and more can come until fulfilment.

I don't fully understand the problem, but all Swaps are unique and must be enforced, we are likely going to add a timestamp, or incremental number to make sure participants applies all Swaps.

Reply:
I think an incremental number is fine, yes. 

To clarify the problem: with a unidirectional channel where the Bob gets more money monotonically: 

```
Voucher 0: (Give $1 to Bob, signed by Alice)
Voucher 1: (Give $2 to Bob, signed by Alice)
```
Voucher `1` is strictly better for Bob than Voucher `0`. Generally Voucher `i` will be better for Alice than Voucher `i-1`. So Alice and Bob can always know which is the "latest voucher" by understanding Bob's incentives. He will always try and use the "latest" (best for him) voucher. For this reason we don't even give the vouchers a number at all.

In a bidirectional channel (of which a currency swap is an example):
```
Voucher 0: (Bob => $1 => Alice && Bob <= £2 <= Alice, signed by Alice and Bob)
Voucher 1: (Bob => $2 => Alice && Bob <= £4 <= Alice, signed by Alice and Bob)
```

Voucher 1 is not obviously "better' for either party than Voucher 1*. Voucher 2 might represent "unswapping" the assets -- kind of natural rebalancing. And we still want to know that Voucher 2 supercedes Voucher 1. So we include the number, since we couldn't infer this just from inspecting the balances / voucher amounts. 

*Although perhaps with the introduction of fees, it would be.